server/src/main/java/org/elasticsearch/action/UnavailableShardsException.java

```diff
@@ -19,41 +19,74 @@
 
 package org.elasticsearch.action;
 
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 import org.elasticsearch.ElasticsearchException;
-import org.jetbrains.annotations.Nullable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.rest.RestStatus;
+import org.jetbrains.annotations.Nullable;
 
-import java.io.IOException;
+import io.crate.exceptions.TableScopeException;
+import io.crate.metadata.RelationName;
+import io.crate.rest.action.HttpErrorStatus;
 
-public class UnavailableShardsException extends ElasticsearchException {
+public class UnavailableShardsException extends ElasticsearchException implements TableScopeException {
+
+    private static final Pattern MSG_INDEX_NAME_PATTERN = Pattern.compile("\\[([^\\]]+)\\].*");
+
+    @Nullable
+    private final RelationName relationName;
 
     public UnavailableShardsException(@Nullable ShardId shardId, String message, Object... args) {
         super(buildMessage(shardId, message), args);
+        this.relationName = shardId == null ? null : RelationName.fromIndexName(shardId.getIndexName());
+    }
+
+    public UnavailableShardsException(ShardId shardId) {
+        super(buildMessage(shardId));
+        this.relationName = shardId == null ? null : RelationName.fromIndexName(shardId.getIndexName());
     }
 
-    public UnavailableShardsException(String index, int shardId, String message, Object... args) {
-        super(buildMessage(index, shardId, message), args);
+    private static String buildMessage(ShardId shardId) {
+        return String.format(Locale.ENGLISH, "[%s] shard %s is not available", shardId.getIndexName(), shardId.id());
     }
 
-    private static String buildMessage(ShardId shardId, String message) {
+    private static String buildMessage(@Nullable ShardId shardId, String message) {
         if (shardId == null) {
             return message;
         }
-        return buildMessage(shardId.getIndexName(), shardId.id(), message);
-    }
-
-    private static String buildMessage(String index, int shardId, String message) {
-        return "[" + index + "][" + shardId + "] " + message;
+        return "[" + shardId.getIndexName() + "][" + shardId.id() + "] " + message;
     }
 
     public UnavailableShardsException(StreamInput in) throws IOException {
         super(in);
+        RelationName name = null;
+        try {
+            Matcher matcher = MSG_INDEX_NAME_PATTERN.matcher(getMessage());
+            name = matcher.matches() ? RelationName.fromIndexName(matcher.group(1)) : null;
```

Comment:
super(in), in turn, uses  ` super(in.readOptionalString(), in.readException());` meaning that we can have NPE on `getMessage()`?

Reply:
That's partly why I swallow the exception - the other reason is that we can have temporary indices which `fromIndexName` can't parse.

Reply:
Another question - are we sure that `getMessage()`, if not null, returns a value in the format, described by the regex? Could it be that it returns an exception than went through wrapping/unwrapping logic?

Reply:
Actually, the message is always not-null, given that it is set in the CTOR. That the base class would also allow the message to be optional has no impact for this concrete implementation.

Even if it weren't, we'd be talking about an exception case for an already exception case. The throw/null check diff here would hardly matter.